### PR TITLE
Update Saber and add AVX2 implementation

### PR DESCRIPTION
**Pull request by jschanck**
New Saber packaging script at https://github.com/jschanck/package-pqclean pulling from https://github.com/KULeuven-COSIC/SABER.

This PR updates the clean implementations and adds AVX2 implementations.

### CODE REVIEW

**Comment by jschanck** [comment]
Occurs to me that I should have packaged the AVX2 code that uses keccak4x. Shouldn't be difficult, but I'm marking as draft for now.

**Comment by jschanck** [comment]
Relevant tests passed.

**Comment by thomwiggers** [comment]
Rebasing on master (pulling in #338) should reduce the number of builds run by quite a bit.

File: crypto_kem/firesaber/avx2/SABER_indcpa.c [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+#include "api.h"
+#include "cbd.h"
+#include "fips202.h"
+#include "pack_unpack.h"
+#include "randombytes.h"
+#include <stdint.h>
+#include <stdio.h>
+#include <string.h>
+//#include "randombytes.h"
+//#include "./polymul/toom_cook_4/toom-cook_4way.c"
```
Comment (thomwiggers) [inline crypto_kem/firesaber/avx2/SABER_indcpa.c:13]: These commented lines can probably go away

File: crypto_kem/firesaber/avx2/SABER_indcpa.c [hunk 2]
```diff
@@ -0,0 +1,416 @@
+#include "./polymul/toom-cook_4way.c"
```
Comment (thomwiggers) [inline crypto_kem/firesaber/avx2/SABER_indcpa.c:1]: ``#include``-ing a `.c` file?

File: crypto_kem/firesaber/avx2/kem.c [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+    randombytes(sk + SABER_SECRETKEYBYTES - SABER_KEYBYTES, SABER_KEYBYTES); // Remaining part of sk contains a pseudo-random number.
+    // This is output when check in PQCLEAN_FIRESABER_AVX2_crypto_kem_dec() fails.
+    return (0);
+}
+
+int PQCLEAN_FIRESABER_AVX2_crypto_kem_enc(uint8_t *c, uint8_t *k, const uint8_t *pk) {
+
+    uint8_t kr[64]; // Will contain key, coins
+    uint8_t buf[64];
```
Comment (thomwiggers) [inline crypto_kem/firesaber/avx2/kem.c:31]: how are the alignment requirements on these buffers... I know the Kyber AVX2 code (on which this code is based) does a bunch of aligning here.

File: crypto_kem/firesaber/avx2/poly.h [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+of "CRYSTALS – Kyber: a CCA-secure module-lattice-based KEM"
+by : Joppe Bos, Leo Ducas, Eike Kiltz, Tancrede Lepoint,
+Vadim Lyubashevsky, John M. Schanck, Peter Schwabe & Damien stehle
+----------------------------------------------------------------------*/
+#include "SABER_params.h"
+#include <stdint.h>
+
+typedef struct {
+    uint16_t coeffs[SABER_N];
+} poly;
```
Comment (thomwiggers) [inline crypto_kem/firesaber/avx2/poly.h:15]: should this be an aligned type?

File: crypto_kem/firesaber/avx2/polymul/toom-cook_4way.c [hunk 5]
```diff
@@ -0,0 +1,1010 @@
...
+
+
+		//------------------AVX evaluation for 1st poly ends------------------
+
+
+		//------------------AVX evaluation for 2nd poly-----------------------
+                    r0_avx=a[small_len_avx];
+                    r1_avx=a[small_len_avx+1];
+                    r2_avx=a[small_len_avx+2];
+                    r3_avx=a[small_len_avx+3];
```
Comment (thomwiggers) [inline crypto_kem/firesaber/avx2/polymul/toom-cook_4way.c:45]: This indentation is weird.

File: crypto_kem/firesaber/avx2/poly.h (hunk 4 above)
Comment (thomwiggers) [reply crypto_kem/firesaber/avx2/poly.h:15]: Probably, I've at least tracked down one place where they get cast to `_m256i`.

**Review by thomwiggers — CHANGES_REQUESTED** [review]
Thanks for the contribution.

The SABER code isn't the greatest; and I've found some things where I thought "hmm". Notably I suspect there are a lot of variables that require explicit alignment...

**Review by thomwiggers — APPROVED** [review]

**Pull request merged**